dmutils/repoutils/freeze-requirements.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
+"""
+Output a file containing the full, frozen dependencies of a requirements file.
+
+This will error out on the pip
+    > Requirement file x contains y, but that package is not installed...
+error caused by `pip install` being case insensitive and the `pip freeze` command being case sensitive.
+
+Usage:
+    freeze-requirements <target>
+    freeze-requirements <target> [options]
+
+Options:
+    <target>             A requirements file for us to freeze
+    --outfile=<outfile>  Output file for frozen requirement values [default: requirements.txt]
+    --verbose
+    -h, --help           Show this screen
+
+Examples:
+    python -m dmutils.repoutils.freeze-requirements requirements-app.txt
+    freeze-requirements.py requirements-app.txt --verbose
+    python -m dmutils.repoutils.freeze-requirements requirements-app.txt --outfile=not-requirements.txt --verbose
+"""
+import logging
+import re
+import shutil
+import subprocess
+import venv
+
+from docopt import docopt
+
+
+logger = logging.getLogger('freeze-requirements')
+
+
+def make_clean(virtualenv_name):
+    try:
+        logger.info(f'Removing {virtualenv_name}')
+        shutil.rmtree(virtualenv_name)
+    except FileNotFoundError:
+        pass
+
+
+def exit_failure(messages, code=1):
+    for message in messages:
+        logger.error(message)
+    exit(code)
+
+
+if __name__ == '__main__':
+    arguments = docopt(__doc__)
+    target = arguments['<target>']
+    outfile = arguments['--outfile']
+    virtualenv_name = 'venv-freeze'
+    logging.basicConfig(level=logging.WARNING if not arguments['--verbose'] else logging.DEBUG)
+
+    make_clean(virtualenv_name)
+
+    logger.info(f'Creating virtualenv: {virtualenv_name}')
+    venv.main(args=(virtualenv_name,))
+
+    logger.info(f'Installing requirements from {target} in virtualenv: {virtualenv_name}')
+    install_cmd = subprocess.run(
+        [f'{virtualenv_name}/bin/pip', 'install', '-r', target],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE
+    )
+    stdout, stderr = str(install_cmd.stdout.decode('utf-8')), str(install_cmd.stderr.decode('utf-8'))
```

Comment:
Sneaky trick, you can use `universal_newlines=True` as an argument to run and then you can get rid of this line 😁 

In Python 3.7 their is a more sensible argument called `text`.

Reply:
```diff
         stderr=subprocess.PIPE
+      universal_newlines=True
     )
-    stdout, stderr = str(install_cmd.stdout.decode('utf-8')), str(install_cmd.stderr.decode('utf-8'))
```

Reply:
Damn, nice subprocess-fu, done 👍

Reply:
You can tell I've been reading through the manual recently 😉

Reply:
Thanks for that , going to merge this now 👍